### PR TITLE
Eric final review and merge, FIP-6 add transfer locked tokens, and integrate get balance changes.

### PR DESCRIPTION
**Pull request by edrotthoff**
This PR adds a new endpoint for transfer locked tokens.
This PR integrates changes to the getFioBalance (addition of available returned)
This PR modifies the tests to test transfer locked tokens, and integrates changes to getFioBalance into the tests.

### CODE REVIEW

File: src/FIOSDK.ts [hunk 1]
```diff
@@ -533,6 +540,35 @@ export class FIOSDK {
...
+   * @param payeePublicKey this is the fio public key for the user to receive locked tokens.
+   * @param canVote true if these locked tokens can be voted, false if these locked tokens are not to be voted.
+   * @param periods this is an array of lockperiods defining the duration and percent of each period, must be in time order.
+   * @param amount this is the amount in SUFs to be transfered.
+   * @param maxFee Maximum amount of SUFs the user is willing to pay for fee. Should be preceded by /get_fee for correct value.
+   * @param technologyProviderId FIO Address of the wallet which generates this transaction.
+   */
+  public transferLockedTokens(
+      payeePublicKey: string,
+      canVote: number,
```
Comment (andreyvEze) [inline src/FIOSDK.ts]: should it be `boolean`?

File: src/transactions/signed/TransferLockedTokens.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import { LockPeriod } from '../../entities/LockPeriod'
+import { validationRules } from '../../utils/validation'
+
+
+export class TransferLockedTokens extends SignedTransaction {
+  public ENDPOINT: string = 'chain/transfer_locked_tokens'
+  public ACTION: string = 'trnsloctoks'
+  public ACCOUNT: string = 'fio.token'
+  public payeePublicKey: string
+  public canVote: number
```
Comment (andreyvEze) [inline src/transactions/signed/TransferLockedTokens.ts:11]: boolean

File: src/transactions/signed/TransferLockedTokens.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    this.validationData = { tpid: technologyProviderId || null }
+    this.validationRules = validationRules.transferLockedTokensRequest
+  }
+
+  public getData(): any {
+    const actor = this.getActor()
+    const data = {
+      payee_public_key: this.payeePublicKey,
+      can_vote: this.canVote,
```
Comment (andreyvEze) [inline src/transactions/signed/TransferLockedTokens.ts:34]: and if here it should be number, could you change it to 
```
can_vote: this.canVote ? 1 : 0,
```

File: src/transactions/signed/TransferLockedTokens.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  constructor(payeePublicKey: string, canVote: number, periods: LockPeriod[], amount: number, maxFee: number, technologyProviderId: string = '') {
+    super()
+    this.payeePublicKey = payeePublicKey
+    this.canVote = canVote
+    this.periods = periods
+    this.amount = amount
+    this.maxFee = maxFee
+    this.technologyProviderId = technologyProviderId
+
+    this.validationData = { tpid: technologyProviderId || null }
```
Comment (andreyvEze) [inline src/transactions/signed/TransferLockedTokens.ts:26]: could we also add `payeePublicKey` here?
```
this.validationData = { tpid: technologyProviderId || null, payeePublicKey }
```

and in validation.ts
```
  transferLockedTokensRequest: {
    tpid: allRules.tpid,
    payeePublicKey: allRules.fioPublicKey,
  },
```

File: src/utils/validation.ts [hunk 5]
```diff
@@ -86,6 +86,9 @@ export const validationRules = {
     tokenCode: allRules.chain,
     tpid: allRules.tpid,
   },
+  transferLockedTokensRequest: {
+    tpid: allRules.tpid,
+  },
```
Comment (andreyvEze) [inline src/utils/validation.ts:91]:
```
  transferLockedTokensRequest: {
    tpid: allRules.tpid,
    payeePublicKey: allRules.fioPublicKey,
  },
```

File: tests/testnet.spec.js [hunk 6]
```diff
@@ -215,7 +215,7 @@ describe('Testing generic actions', () => {
   it(`getFioBalance`, async () => {
     const result = await fioSdk.genericAction('getFioBalance', {})
 
-    expect(result).to.have.all.keys('balance')
+    expect(result).to.have.all.keys('balance','available')
     expect(result.balance).to.be.a('number')
```
Comment (andreyvEze) [inline tests/testnet.spec.js:222]: add `expect(result.available).to.be.a('number')`

File: tests/testnet.spec.js [hunk 7]
```diff
@@ -502,7 +507,7 @@ describe('Testing generic actions', () => {
       fioPublicKey: publicKey2
     })
 
-    expect(result).to.have.all.keys('balance')
+    expect(result).to.have.all.keys('balance','available')
     expect(result.balance).to.be.a('number')
```
Comment (andreyvEze) [inline tests/testnet.spec.js:510]: add `expect(result.available).to.be.a('number')`

File: tests/testnet.spec.js [hunk 8]
```diff
@@ -438,6 +438,11 @@ describe('Testing generic actions', () => {
           chain_code: ethChainCode,
           token_code: ethTokenCode,
           public_address: 'xxxxxxyyyyyyzzzzzz1',
+        },
+        {
+          chain_code: fioChainCode,
+          token_code: fioTokenCode,
+          public_address: publicKey,
```
Comment (andreyvEze) [inline tests/testnet.spec.js]: do we need it here?

File: src/transactions/signed/TransferLockedTokens.ts (hunk 4 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:26]: I actually had that, and that validator is failing, will leave it for later if thats ok.

File: tests/testnet.spec.js [hunk 9]
```diff
@@ -747,11 +752,67 @@ describe('Transfer tokens', () => {
...
+  })
+
+  it(`Transfer locked tokens`, async () => {
+    try {
+      var mnem = FIOSDK.getMnemonic()
+      var privKey = await FIOSDK.createPrivateKeyMnemonic(mnem)
+      let pubkey =  FIOSDK.derivedPublicKey(privKey.fioKey)
+      const result = await fioSdk.genericAction('transferLockedTokens', {
+        payeePublicKey: pubkey.publicKey,
+        canVote: 0,
```
Comment (andreyvEze) [inline tests/testnet.spec.js]: change to `false`

File: src/utils/validation.ts (hunk 5 above)
Comment (edrotthoff) [reply src/utils/validation.ts:91]: same comment the fioPublicKey validator isn't working presently, will go back and check it out..

File: tests/testnet.spec.js [hunk 10]
```diff
@@ -747,11 +752,67 @@ describe('Transfer tokens', () => {
...
+    await timeout(10000)
+    const result = await fioSdk2.genericAction('getFioBalance', {})
+    fioBalanceAfter = result.balance
+    expect(fundsAmount).to.equal(1000000000)
+    }catch (e) {
+      console.log(e);
+    }
+  })
+})
+
```
Comment (andreyvEze) [inline tests/testnet.spec.js:815]: would you add the same changes to the `index.spec.js`?

File: src/transactions/signed/TransferLockedTokens.ts (hunk 4 above)
Comment (andreyvEze) [reply src/transactions/signed/TransferLockedTokens.ts:26]: hm, ok, let's do not add this, we need to check why it is failing

File: src/utils/validation.ts (hunk 5 above)
Comment (andreyvEze) [reply src/utils/validation.ts:91]: let's leave as was before for now 
```
 transferLockedTokensRequest: {
    tpid: allRules.tpid,
  },
```

File: src/FIOSDK.ts (hunk 1 above)
Comment (edrotthoff) [reply src/FIOSDK.ts]: this was changed

File: src/transactions/signed/TransferLockedTokens.ts (hunk 2 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:11]: done

File: src/transactions/signed/TransferLockedTokens.ts (hunk 4 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:26]: we will complete this work in a follow on story.

File: src/utils/validation.ts (hunk 5 above)
Comment (edrotthoff) [reply src/utils/validation.ts:91]: we will resolve this in a follow on story

File: tests/testnet.spec.js (hunk 6 above)
Comment (edrotthoff) [reply tests/testnet.spec.js:222]: done

File: tests/testnet.spec.js (hunk 8 above)
Comment (edrotthoff) [reply tests/testnet.spec.js]: done

File: tests/testnet.spec.js (hunk 9 above)
Comment (edrotthoff) [reply tests/testnet.spec.js]: done

File: tests/testnet.spec.js (hunk 7 above)
Comment (edrotthoff) [reply tests/testnet.spec.js:510]: done

File: src/transactions/signed/TransferLockedTokens.ts (hunk 3 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:34]: done

File: src/transactions/signed/TransferLockedTokens.ts (hunk 2 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:11]: did it just little different, but the signature is boolean now

File: src/transactions/signed/TransferLockedTokens.ts (hunk 4 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:26]: yes, will do it in a future story

File: src/transactions/signed/TransferLockedTokens.ts (hunk 3 above)
Comment (edrotthoff) [reply src/transactions/signed/TransferLockedTokens.ts:34]: did it just a bit different, but its boolean in the signature

File: src/utils/validation.ts (hunk 5 above)
Comment (edrotthoff) [reply src/utils/validation.ts:91]: we will do it in a future story.

File: tests/testnet.spec.js (hunk 8 above)
Comment (edrotthoff) [reply tests/testnet.spec.js]: removed

File: tests/testnet.spec.js (hunk 10 above)
Comment (edrotthoff) [reply tests/testnet.spec.js:815]: yes, doing it.

**Comment by edrotthoff** [comment]
closed this pull request, the dev branch will no longer integrate with develop, made a new branch and placed changes into this branch.

**Pull request closed without merging**